### PR TITLE
fix bytes.setUInt16 in IE8/9

### PR DESCRIPTION
**Pull request by R32**
```hx
class Main {
    static function main() {
        var b = haxe.io.Bytes.alloc(128);
        var v = 101;
        b.setUInt16(0, v);
        trace(b.get(0) == v);
        trace(b.getUInt16(0) == v);
        trace(b.getInt32(0) == v);
    }
}
```

### CODE REVIEW

**Comment by R32** [comment]
I want to move `i32ToFloat/FloatToi32/i64ToDouble/doubleToI64 ` into **`js.html.compat.DataView`** to prevent infinite recursion. But I found that the pure calculation in **`haxe.io.FPHelper`** is not correct!

**Comment by RealyUniqueName** [comment]
Will it fix #6685?

**Comment by R32** [comment]
It works very well in IE8.  
```haxe
class Main {
    static function main() {
        var b = haxe.io.Bytes.alloc(8);
        b.setDouble(0, 1.1976931348623157e+308);
        trace(b.getInt32(0) == 0x493F546C);
        trace(b.getInt32(4) == 0x7FE551D4);

        b.setFloat(0, -3.3360025354124689e-013);
        trace(b.getInt32(0) == 0xAABBCCDD);

        b.setInt32(0, 0xAABBCCDD);
        trace(b.getFloat(0) == -3.3360025354124689e-013);

        b.setInt32(0, 0x493F546C);
        b.setInt32(4, 0x7FE551D4);
        trace(b.getDouble(0) == 1.1976931348623157e+308);
    }
}
```
Here is  [uint-test-pr6697.zip](https://github.com/HaxeFoundation/haxe/files/1407313/uint-test-pr6697.zip)

**Comment by Simn** [comment]
Shouldn't we fix FPHelper instead of bypassing it?

**Comment by R32** [comment]
I think we can only bypass FPHelper, because it will fall into infinite recursion.

**Comment by ncannasse** [comment]
That seems good to me. If we want to factorize the code, that would require putting FP emulation function of FPHelper in separate functions (still in FPHelper.hx) so they can be called either by either other FPHelper functions or by DataView emulation.

**Comment by R32** [comment]
unit test
```hx
import haxe.io.FPHelper;
import StringTools.hex;

class Main {
    static function main() {
        var b = haxe.io.Bytes.alloc(8);
        b.setDouble(0, 1.1976931348623157e+308);
        eq(b.getInt32(0) == 0x493F546C);
        eq(b.getInt32(4) == 0x7FE551D4);

        b.setFloat(0, -3.3360025354124689e-013);
        eq(b.getInt32(0) == 0xAABBCCDD);
        b.setInt32(0, 0xAABBCCDD);
        eq(b.getFloat(0) == -3.3360025354124689e-013);

        b.setInt32(0, 0x493F546C);
        b.setInt32(4, 0x7FE551D4);
        eq(b.getDouble(0) == 1.1976931348623157e+308);

        i2f(0x006CFFCF);
        i2f(0xAABBCCDD);
        i2f(0xFF800001);               // Math.NaN
        i2f(0xFF800000);               // Math.NEGATIVE_INFINITY
        i2f(0x7F800000);               // Math.POSITIVE_INFINITY
        i2f(0x7F800001);               // Math.NaN
        i2f(0x7F7FFFFF);               // MAX_VALUE float32
        i2f(1);                        // MIN_VALUE float32
        i2f(0xFF7FFFFF);
        i2f(); i2f(); i2f(); i2f(); i2f();
        i2f(); i2f(); i2f(); i2f(); i2f();
        i2f(); i2f(); i2f(); i2f(); i2f();
        i2f(); i2f(); i2f(); i2f(); i2f();
        i2f(); i2f(); i2f(); i2f(); i2f();

        ii2d(0, 0xFFF00000);           // Math.NEGATIVE_INFINITY
        ii2d(0, 0x7FF00000);           // Math.POSITIVE_INFINITY
        ii2d(1, 0xFFF00000);           // Math.NaN
        ii2d(1, 0x7FF00000);           // Math.NaN
        ii2d(0xFFFFFFFF, 0x7FEFFFFF);  // MAX_VALUE
        ii2d(1, 0);                    // MIN_VALUE
        ii2d(0xAABBCCDD, 0xAABBCCDD);
        ii2d(0xF96F73E0 , 0xFFFFF);
        ii2d(0xC3940728, 0xFFFFF);
        ii2d(0xFFFFFFFF, 0xFFFFF);
        ii2d(0x8FF00000, 0xFFFFF);
        ii2d(null, 0xFFFFF);
        ii2d(null, 0xFFFFF);
        ii2d(); ii2d(); ii2d(); ii2d(); ii2d();
        ii2d(); ii2d(); ii2d(); ii2d(); ii2d();
        ii2d(); ii2d(); ii2d(); ii2d(); ii2d();
        ii2d(); ii2d(); ii2d(); ii2d(); ii2d();
        ii2d(); ii2d(); ii2d(); ii2d(); ii2d();

        f2i(-3.3360025354124689e-013); // 0xAABBCCDD
        f2i(2.3479298072268004e-038);  // 0x00FFAABB (exp == -126)
        f2i(1.001000041759629e-38);    // 0x006CFFCF (exp == -127)
        f2i(1.793662034335766e-43);
        f2i(1.8751230111111);
        f2i(3.190147189883798e+38);    // 0x7F700000 (exp == 127)
        f2i(3.4028234663852886e+38);   // 0x7F7FFFFF (exp == 127)
        f2i(3.4028334663852886e+38);   // 3.4028234663852886e+38 + 0.00001 (exp == 128)
        f2i(-3.4028234663852886e+38);  // 0xFF7FFFFF (exp == 127)
        f2i(); f2i(); f2i(); f2i(); f2i();
        f2i(); f2i(); f2i(); f2i(); f2i();
        f2i(); f2i(); f2i(); f2i(); f2i();
        f2i(); f2i(); f2i(); f2i(); f2i();
        f2i(); f2i(); f2i(); f2i(); f2i();
        f2i(); f2i(); f2i(); f2i(); f2i();

        #if !(hl || cs || java || neko)// These platforms: "Std.int" can not handle big numbers correctly.
        d2ii(1.7976931348623157e+308); // 0x7FEF FFFF FFFF FFFF
        d2ii(5e-324);                  // 0x0000 0000 0000 0001
        d2ii(2.2250729296159775e-308); // 0x000F FFFF 8FF0 0000
        d2ii(1.1976931348623157e+308); // 0x7FE5 51D4 493F 546C
        d2ii(); d2ii(); d2ii(); d2ii(); d2ii();
        d2ii(); d2ii(); d2ii(); d2ii(); d2ii();
        d2ii(); d2ii(); d2ii(); d2ii(); d2ii();
        d2ii(); d2ii(); d2ii(); d2ii(); d2ii();
        d2ii(); d2ii(); d2ii(); d2ii(); d2ii();
        d2ii(); d2ii(); d2ii(); d2ii(); d2ii();
        d2ii(); d2ii(); d2ii(); d2ii(); d2ii();
        #end
        trace("done!");
    }

    static function eq(b: Bool) if (!b) throw "TODO";

    static function fp_i2f(i) return @:privateAccess FPHelper.i2f(i);
    static function fp_f2i(f) return @:privateAccess FPHelper.f2i(f);
    static function fp_ii2d(lo, hi) return @:privateAccess FPHelper.ii2d(lo, hi);
    static function fp_d2ii(d)      return @:privateAccess FPHelper.d2ii(d);

    macro static function i2f(?i: Int) {
        if (i == null) i = Std.int(Math.random() * 0x7FFFFFFF);
        var si = "0x" + hex(i, 8);
        var b = haxe.io.Bytes.alloc(8);
        b.setInt32(0, i);
        var f = b.getFloat(0);

        return macro {
            var r = fp_i2f($v{i});
            //trace("i2f(" + $v{si} +"): " + r + " == " +$v{f});
            if (Math.isNaN($v{f}))
                eq(Math.isNaN(r));
            else
                eq(r == $v{f});
        }
    }

    macro static function ii2d(?lo: Int, ?hi: Int) {
        if (lo == null) lo = Std.int(Math.random() * 0x7FFFFFFF);
        if (hi == null) hi = Std.int(Math.random() * 0x7FFFFFFF);
        var b = haxe.io.Bytes.alloc(8);
        b.setInt32(0, lo);
        b.setInt32(4, hi);
        var f = b.getDouble(0);

        return macro {
            var r = fp_ii2d($v{ lo }, $v{ hi } );
            //trace("ii2d(0x" + $v{ hex(lo) } +", 0x" + $v{ hex(hi) } + "): " + r + " == " + $v{f});
            if (Math.isNaN($v{f}))
                eq(Math.isNaN(r))
            else
                eq(r == $v{f});
        }
    }

    macro static function f2i(?f: Float) {
        var b = haxe.io.Bytes.alloc(8);
        if (f == null) {
            while (true) {
                b.setInt32(0, Std.int(Math.random() * 0x7FFFFFFF));
                f = b.getFloat(0);
                if (!Math.isNaN(f)) break;
            }
        } else {
            b.setFloat(0, f);
        }
        var n = b.getInt32(0);
        return macro {
            var r = fp_f2i($v{f});
            //trace(hex(r) + " == " +hex($v{n}) + " == " + $v{f});
            eq(r == $v{n});
        }
    }

    macro static function d2ii(?df: Float) {
        var b = haxe.io.Bytes.alloc(8);
        if (df == null) {
            while (true) {
                b.setInt32(0, Std.int(Math.random() * 0x7FFFFFFF));
                b.setInt32(4, Std.int(Math.random() * 0x7FFFFFFF));
                df = b.getDouble(0);
                if (!Math.isNaN(df)) break;
            }
        } else {
            b.setDouble(0, df);
        }
        var lo = b.getInt32(0);
        var hi = b.getInt32(4);
        return macro {
            var r = fp_d2ii($v{ df } );
            //trace(
                //"(0x" + hex(r.low,  8) + ", 0x" + hex(r.high,   8) + ") == (0x"
                      //+ hex($v{lo},8) + ", 0x" + hex($v{hi}, 8) + ")"
            //);
            eq(r.low == $v{lo} && r.high == $v{hi});
        }
    }
}
```

File: std/haxe/io/FPHelper.hx [hunk 1]
```diff
@@ -56,7 +56,7 @@ class FPHelper {
 		static var helper = new js.html.DataView(new js.html.ArrayBuffer(8));
 	#end
 
-	#if !(neko || cpp || cs || java || flash || nodejs)
+	#if !(neko || cpp || cs || java || flash || (js && nodejs))
```
Comment (Gama11) [inline std/haxe/io/FPHelper.hx]: This shouldn't be necessary with https://github.com/HaxeFoundation/hxnodejs/pull/94?

Comment (R32) [reply std/haxe/io/FPHelper.hx]: Yes, but it makes CI be green, and some people may not updated their hxnodejs

File: std/haxe/io/FPHelper.hx [hunk 2]
```diff
@@ -54,8 +54,83 @@ class FPHelper {
...
+			var e = (i >> 23) & 0xff;
+			if (e == 255)
+				return i & 0x7fffff == 0
+					? (sign > 0 ? Math.POSITIVE_INFINITY : Math.NEGATIVE_INFINITY)
+					: Math.NaN;
+			var m = e == 0 ? (i & 0x7fffff) << 1 : (i & 0x7fffff) | 0x800000;
+			return sign * m * Math.pow(2, e - 150);
+		}
+
+		static inline function ii2d(lo: Int, hi: Int): Float {
```
Comment (RealyUniqueName) [inline std/haxe/io/FPHelper.hx]: Please, rename these methods to something human-readable

Comment (R32) [reply std/haxe/io/FPHelper.hx]: So, which names are appropriate? :sweat_smile: `i64_to_double/double_to_i64`, `i32_to_float/float_to_i32` or something else?

Comment (RealyUniqueName) [reply std/haxe/io/FPHelper.hx]: Like that, but with camel case )

Comment (R32) [reply std/haxe/io/FPHelper.hx]: These names `i32ToFloat/floatToI32/i64ToDouble/doubleToI64` are already exists,  I can only get the following:
```
i2f, f2i, i2f64, f2i64,

itof32, ftoi32, itof64, ftoi64,
		
i32ToF32, f32ToI32, i64ToF64, f64ToI64

_i32ToFloat, _floatToi32, _i64ToDouble, _doubleToI64
```
Any suggestions?

Comment (RealyUniqueName) [reply std/haxe/io/FPHelper.hx]: The latter option looks best to me.

**Comment by R32** [comment]
[unit-test-pr6697.zip](https://github.com/HaxeFoundation/haxe/files/1423118/unit-test-pr6697.zip)

**Comment by R32** [comment]
The travis-ci did not succeed. But it doesn't look like my fault.

**Comment by andyli** [comment]
@ncannasse Would you review and merge it?

**Comment by ncannasse** [comment]
I have fixed the specic setUint16 issue in 2b3136724e808c41c07cb3d6bcd10845ba6db191

**Comment by ncannasse** [comment]
Also fixed the setFloat64. I want to merge this but I'm worried about this big #if that lists too many targets. Can't we have inclusive targets instead of !(almost every target)

**Comment by R32** [comment]
I have rebase it and have tried to improve the big conditions.

**Comment by farteryhr** [comment]
a test with power-of-twos and their neighbors.
https://try.haxe.org/#3abF0
this emulated conversion still needs more love...
on 4.0.0 preview 4, forcing to use polyfill (just with 4.0.0 p3 it seems to be still infinite recursion... better to have specification on where to do the real job...) and ran on nodejs.

btw i suggest to rename i64ToDouble to like i64AsDouble, as it does reinterpretation, the "to" sounds like conversion preserving the represented value.

btw2 is there convention whether to name them float/double, single/double or float32/float64?

---
my fix:
```haxe
	static inline var F32INF = 0x7F800000;

	static inline function _floatToI32(f: Float): Int {
		if (f == 0) return 0;
		if (Math.isNaN(f)) return 0x7FFFFFFF;
		var sign =  f > 0 ? 0 : 0x80000000;
		if (!Math.isFinite(f)) return F32INF | sign;
		var af = f < 0 ? -f : f;
		var exp = Math.floor(Math.log(af) / LN2); // needs correction
		var epot = Math.pow(2, exp); // always precise
		if (af / epot >= 2) {
			exp++;
			epot *= 2;
		} else if(af / epot < 1) {
			exp--;
			epot *= 0.5;
		}
		if (exp > 127) {
			return F32INF | sign;
		} else {
			if (exp <= -127 ) {
				exp = -127;
				af *= 7.1362384635298e+44;  // af * 0.5 * 0x800000 / Math.pow(2, -127)
			} else {
				af = (af / epot - 1.0) * 0x800000;
			}
			return sign | ((exp + 127) << 23) | Math.round(af);
		}
	}

	static inline function _doubleToI64(v: Float): Int64 @:privateAccess {
		var i64 = i64tmp;
		if( v == 0 ) {
			i64.set_low(0);
			i64.set_high(0);
		} else if (Math.isNaN(v)) {
			i64.set_low(0xFFFFFFFF);
			i64.set_high(0x7FFFFFFF);
		} else if (!Math.isFinite(v)) {
			i64.set_low(0);
			i64.set_high(v > 0 ? 0x7FF00000 : 0xFFF00000);
		} else {
			var av = v < 0 ? -v : v;
			var exp = Math.floor(Math.log(av) / LN2); // needs correction
			if (exp >= 1024)
				exp = 1023; // for max double, exp == 1024 (by error), causes epot = Infinity
			var epot = Math.pow(2, exp); // always precise
			if (av / epot >= 2) {
				exp++;
				epot *= 2;
			} else if(av / epot < 1) {
				exp--;
				epot *= 0.5;
			}
			if (exp <= -1023) {
				exp = -1023;
				av = av / 2.2250738585072014e-308;
			} else {
				av = av / epot - 1.0;
			}
			var sig = Math.fround(av * 4503599627370496.); // 2^52
			// Note: If "sig" is outside of the signed Int32 range, the result is unspecified in HL, C#, Java and Neko,
			var sig_l = Std.int(sig);
			var sig_h = Std.int(sig / 4294967296.0);
			i64.set_low(sig_l);
			i64.set_high((v < 0 ? 0x80000000 : 0) | ((exp + 1023) << 20) | sig_h);
		}
		return i64;
	}
```

**Pull request merged**